cpp/src/endpoint.cpp

```diff
@@ -120,7 +124,9 @@ std::shared_ptr<Endpoint> createEndpointFromHostname(std::shared_ptr<Worker> wor
   params.sockaddr.addrlen = info->ai_addrlen;
   params.sockaddr.addr    = info->ai_addr;
 
-  return std::shared_ptr<Endpoint>(new Endpoint(worker, &params, endpointErrorHandling));
+  auto ep = std::shared_ptr<Endpoint>(new Endpoint(worker, &params, endpointErrorHandling));
+  ep->_callbackData->endpoint = ep;
+  return ep;
```

Comment:
With my suggestion above, perhaps this would then be:

```
auto ep = std::shared_ptr<Endpoint>(new Endpoint(worker, &params, endpointErrorHandling));
ep.setCallback(...);
```

But, I realise this has a problem, because we pass `params` into `ucp_ep_create` and if we only set the error callback later, then it won't be hooked up.

Should this then be a three-phase operation:

```
auto ep = std::make_shared<Endpoint>(new Endpoint(worker, &params, endpointErrorHandling));
ep.setCallback(...);
ep.create();
```
WDYT?

It's a bigger refactor than we have here though, and still leaves the option of forgetting one of the steps. So perhaps it's not worth it now.

Reply:
This is certainly not what I would like to have but I agree with you it's a better implementation. We can even avoid the need for `setCallback()`, a `create()` method is sufficient. See https://github.com/rapidsai/ucxx/pull/227/commits/b36ed91c18bc883d5c9ad13b09a374aab2a595bb .